llvm/lib/CodeGen/WindowScheduler.cpp

```diff
@@ -437,12 +437,16 @@ int WindowScheduler::calculateMaxCycle(ScheduleDAGInstrs &DAG,
       int PredCycle = getOriCycle(PredMI);
       ExpectCycle = std::max(ExpectCycle, PredCycle + (int)Pred.getLatency());
     }
-    // ResourceManager can be used to detect resource conflicts between the
-    // current MI and the previously inserted MIs.
-    while (!RM.canReserveResources(*SU, CurCycle) || CurCycle < ExpectCycle) {
-      ++CurCycle;
-      if (CurCycle == (int)WindowIILimit)
-        return CurCycle;
+    // Zero cost instructions do not need to check resource.
+    if (!TII->isZeroCost(MI.getOpcode())) {
```

Comment:
Why does this hook exist? How is it different from isMetaInstruction?

Reply:
Also copy should need resource checks?

Reply:
> Why does this hook exist? How is it different from isMetaInstruction?
Metainstructions do not participate in scheduling, but zero-cost instructions do. `COPY` is a typical example of a zero-cost instruction

> Also copy should need resource checks?
`COPY` do not occupy hardware resources so they do not require resource checks, they still need to be handled during instruction scheduling to ensure proper execution.

Reply:
> COPY do not occupy hardware resources 

This is a very target specific assumption but ok